src/server/frontend_wayland/fractional_scale_v1.cpp

```diff
@@ -0,0 +1,105 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "fractional_scale_v1.h"
+
+#include "mir/wayland/protocol_error.h"
+#include "unique_resource_tracker.h"
+#include "wl_surface.h"
+
+#include <boost/throw_exception.hpp>
+#include <memory>
+
+namespace mir
+{
+namespace frontend
+{
+class FractionalScaleManagerV1: public wayland::FractionalScaleManagerV1
+{
+    public:
+        FractionalScaleManagerV1(struct wl_resource* resource);
+
+        class Global: public wayland::FractionalScaleManagerV1::Global
+        {
+            public:
+                Global(wl_display* display);
+
+            private:
+                virtual void bind(wl_resource* new_wp_fractional_scale_manager_v1) override;
+        };
+
+    private:
+        virtual void get_fractional_scale(struct wl_resource* id, struct wl_resource* surface) override;
+        std::shared_ptr<UniqueResourceTracker> surfaces_with_fractional_scale;
+};
+
+class FractionalScaleV1: public wayland::FractionalScaleV1
+{
+    public:
+        FractionalScaleV1(struct wl_resource* resource);
+};
+}
+}
+
+namespace mf = mir::frontend;
+
+auto mf::create_fractional_scale_v1(wl_display* display)
+    -> std::shared_ptr<wayland::FractionalScaleManagerV1::Global>
+{
+    return std::make_shared<FractionalScaleManagerV1::Global>(display);
+}
+
+mf::FractionalScaleManagerV1::FractionalScaleManagerV1(struct wl_resource* resource)
+    : mir::wayland::FractionalScaleManagerV1{resource, Version<1>{}},
+      surfaces_with_fractional_scale{std::make_shared<UniqueResourceTracker>()}
+{
+}
+
+mf::FractionalScaleManagerV1::Global::Global(wl_display* display)
+    : wayland::FractionalScaleManagerV1::Global{display, Version<1>{}}
+{
+}
+
+void mf::FractionalScaleManagerV1::Global::bind(wl_resource* new_wp_fractional_scale_manager_v1)
+{
+    new FractionalScaleManagerV1{new_wp_fractional_scale_manager_v1};
+}
+
+void mf::FractionalScaleManagerV1::get_fractional_scale(struct wl_resource* id, struct wl_resource* surface)
+{
+    if (!surfaces_with_fractional_scale->register_resource_owner(surface))
+    {
+        BOOST_THROW_EXCEPTION(mir::wayland::ProtocolError(
+            resource, Error::fractional_scale_exists, "Surface already has a fractional scale component attached"));
+    }
+
+    const auto scale = 1.5f;
+    auto fractional_scale = new FractionalScaleV1{id};
+    fractional_scale->send_preferred_scale_event(120 * scale);
+
+    auto surf = WlSurface::from(surface);
+    auto on_destroy = [this, surface]()
+    {
+        surfaces_with_fractional_scale->unregister_resource_owner(surface);
+    };
+    surf->add_destroy_listener(on_destroy);
+    fractional_scale->add_destroy_listener(on_destroy);
```

Comment:
Hm. I suspect this is the wrong approach. I think it's likely that a full implementation of this will want to store the `FractionalScaleV1` in the surface - we'll need to send out a new `prefered_scale` whenever the surface moves to an output with different scale.

(The [`wp_viewporter` merge](https://github.com/canonical/mir/pull/3445/files) might give some pointers as to how that would go, although because the fractional-scale object isn't double-buffered state you wouldn't need the pending/current split)

Reply:
Should be good now! Thanks for all the help today :)